test/unit/test-cookies.js

```diff
@@ -14,142 +14,215 @@
-          },
-        },
-        cookieName
-      )
-    );
-  }
+import * as lolex from 'lolex';
+import {BASE_CID_MAX_AGE_MILLIS} from '../../src/service/cid-impl';
+import {
+  getCookie,
+  getHighestAvailableDomain,
+  setCookie,
+} from '../../src/cookies';
+
+describes.fakeWin('test-cookies', {amp: true}, env => {
+  let win;
+  let clock;
+  let doc;
+
+  beforeEach(() => {
+    win = env.win;
+    doc = win.document;
+    clock = lolex.install({
+      target: window,
+      now: new Date('2018-01-01T08:00:00Z'),
+    });
+  });
+
+  afterEach(() => {
+    clock.uninstall();
+  });
 
   it('should return null for no cookie, malformed, or not found', () => {
-    expectCookie(null, 'c1').to.be.null;
-    expectCookie(undefined, 'c1').to.be.null;
-    expectCookie('', 'c1').to.be.null;
-    expectCookie('c1', 'c1').to.be.null;
-    expectCookie('e2=1', 'c1').to.be.null;
+    expect(doc.cookie).to.equal('');
+    expect(getCookie(win, 'c1')).to.be.null;
+    expect(getCookie(win, 'c1')).to.be.null;
+    doc.cookie = 'c1';
+    expect(getCookie(win, 'c1')).to.be.null;
+    doc.cookie = 'e2=1';
+    expect(getCookie(win, 'c1')).to.be.null;
   });
 
   it('should return value when found', () => {
-    expectCookie('c1=1', 'c1').to.equal('1');
-    expectCookie(' c1 = 1 ', 'c1').to.equal('1');
-    expectCookie('c1=1;c2=2', 'c1').to.equal('1');
-    expectCookie('c1=1;c2=2', 'c2').to.equal('2');
+    doc.cookie = 'c1=1';
+    expect(getCookie(win, 'c1')).to.equal('1');
+    doc.cookie = ' c2 = 2 ';
+    expect(doc.cookie).to.equal('c1=1; c2 = 2 ');
+    expect(getCookie(win, 'c1')).to.equal('1');
+    expect(getCookie(win, 'c2')).to.equal('2');
   });
 
   it('should return value for an escaped cookie name', () => {
-    expectCookie('c%26=1', 'c&').to.equal('1');
+    doc.cookie = 'c%26=1';
+    expect(doc.cookie).to.equal('c%26=1');
+    expect(getCookie(win, 'c&')).to.equal('1');
   });
 
   it('should return an unescaped value', () => {
-    expectCookie('c1=1%26', 'c1').to.equal('1&');
+    doc.cookie = 'c1=1%26';
+    expect(doc.cookie).to.equal('c1=1%26');
+    expect(getCookie(win, 'c1')).to.equal('1&');
   });
 
   it('should write the cookie', () => {
-    const doc = {};
-    setCookie(
-      {
-        document: doc,
-        location: {
-          href: 'https://www.example.com/test',
-        },
-      },
-      'c&1',
-      'v&1',
-      1447383159853
-    );
-    expect(doc.cookie).to.equal(
-      'c%261=v%261; path=/; expires=Fri, 13 Nov 2015 02:52:39 GMT'
+    setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS);
+    expect(doc.lastSetCookieRaw).to.equal(
+      'c%261=v%261; path=/; expires=Tue, 01 Jan 2019 08:00:00 GMT'
     );
+    expect(doc.cookie).to.equal('c%261=v%261');
+  });
+
+  it('getHigestAvailableDomain without meta tag', () => {
+    // Proxy Origin
+    win.location.href = 'https://foo-bar.cdn.ampproject.org/c/foo.bar.com';
+    expect(getHighestAvailableDomain(win)).to.be.null;
```

Comment:
why this is null. this should be `foo.bar.com`

Reply:
with current logic `https://example.com/` will have different HighestAvailableDomain on proxy vs origin

Reply:
Unfortunately, it would very difficult to calculate etld+1 on proxy origin (load and parse 30+ kB of data for every user) . And that's the reason why we ask cache to provide the `<meta name='amp-scope-coolie>` in the first place. 
Given the function will only be used for writing cookie (should not happen in proxy origin). Calculating subdomains for linker (has `<meta name='amp-scope-cookie'>` added by cache). It won't be an issue, and it's some compromise we decided to make. 

I could add a comment on that to clarify.